python/ray/ml/config.py

```diff
@@ -35,29 +34,71 @@ class ScalingConfigDataClass:
         Strategies <pgroup-strategy>` for the possible options.
     """
 
-    num_workers: int
+    num_workers: int = 1
     use_gpu: bool = False
     resources_per_worker: Optional[Dict] = None
     placement_strategy: str = "PACK"
 
+    def __post_init__(self):
+        self.resources_per_worker = (
+            self.resources_per_worker if self.resources_per_worker else {}
+        )
+
+        if self.num_workers < 0:
+            raise ValueError("`num_workers` must be a positive integer.")
+
+        if self.resources_per_worker:
+            # Override CPU and GPU resources and remove from dict.
+            if not self.use_gpu and self.num_gpus_per_worker > 0:
+                raise ValueError(
+                    "`use_gpu` is False but `GPU` was found in "
+                    "`resources_per_worker`. Either set `use_gpu` to True or "
+                    "remove `GPU` from `resources_per_worker."
+                )
+
+            if self.use_gpu and self.num_gpus_per_worker == 0:
+                raise ValueError(
+                    "`use_gpu` is True but `GPU` is set to 0 in "
+                    "`resources_per_worker`. Either set `use_gpu` to False or "
+                    "request a positive number of `GPU` in "
+                    "`resources_per_worker."
+                )
+
     @property
     def num_cpus_per_worker(self):
         """The number of CPUs to set per worker."""
-        raise NotImplementedError
+        return self.resources_per_worker.get("CPU", 1)
 
     @property
     def num_gpus_per_worker(self):
         """The number of GPUs to set per worker."""
-        raise NotImplementedError
+        return self.resources_per_worker.get("GPU", int(self.use_gpu))
 
     @property
     def additional_resources_per_worker(self):
         """Resources per worker, not including CPU or GPU resources."""
-        raise NotImplementedError
+        return {
+            k: v
+            for k, v in self.resources_per_worker.items()
+            if k not in ["CPU", "GPU"]
+        }
 
-    def as_placement_group_factory(self) -> "PlacementGroupFactory":
+    def as_placement_group_factory(self) -> PlacementGroupFactory:
         """Returns a PlacementGroupFactory to specify resources for Tune."""
-        raise NotImplementedError
+        trainer_bundle = [{"CPU": 0}]
```

Comment:
Why 0 and not 1 by default?

Reply:
More generally, should we add config for the trainer (like in RLlib)? Fine if we decide to punt this for later.

Reply:
This is from the suggestion from tune team. It seems like using 0 should not be a problem cc @xwjiang2010 @Yard1 @krfricke? 

The benefit of having 0 is that the resources model is exactly like what's in scaling config. The num cpus required is equal to num_workers, and users don't have to account for the extra cpu.

Reply:
With Train today this has been a source of confusion and users have run into problems underspecifying resources.

Maybe we can extend `ScalingConfig` to include resources for the `Trainer` itself. Since this is executed via Ray Tune, we always know that the `Trainer` will be executed remotely (compared to the existing Ray Train, where we had to separate the `BackendExecutor`).

Reply:
@matthewdeng I don't think that fully resolves the issue though. Even if we automatically include it in scaling config, there is still an extra CPU that is required that the user is not aware of. 

So if you have a node with 4 CPUs, then you won't be able to do training with 4 workers, and it won't be clear why.

Assuming that using 0 CPUs has no negative impact (which is what my current understanding is), then I'd vote to keep as is. But let me check again with Tune team to confirm.

Reply:
Oh, I meant if the `ScalingConfig` could be more generically thought of a "resource config" that contains resources for each worker, plus the resources for the `Trainer` itself. For the single-threaded trainer case, the user still needs to specify resources somehow right?

Reply:
Got it, ok yeah this is @ericl's suggestion of adding a config for the trainer.

Any suggestions for the API as well as the default value?

Just have `num_cpus_for_driver`?

Reply:
To me, the ideal situation would be to use 0 cpus if not running with Tuner vs. 1 cpu if you are. 

This is also the same as what rllib is doing 
```
# Number of CPUs to allocate for the trainer. Note: this only takes effect
# when running in Tune. Otherwise, the trainer runs in the main program.
"num_cpus_for_driver": 1,
```

Reply:
IMO you can just have a `trainer_resources` dict, and use that to override the default values. This should provide parity with the resource configuration we provide for each worker in the distributed case, and easily translate to the `PlacementGroupFactory`.

With the new implementation, shouldn't it always be executed through Tuner?

Reply:
Ok added trainer_resources dict that defaults to {"CPU": 1}